### ISSUE TITLE
Dreambooth broken, possibly because of ADAM optimizer, possibly more.

### ISSUE THREAD

**Issue by affableroots** [huggingface/diffusers#712]
I think Huggingface's Dreambooth is the only popular SD implementation that also uses Prior Preservation Loss, so I've been motivated to get it working, but the results have been terrible, and the entire model degrades, regardless of: # timesteps, learning rate, PPL turned on/off, # instance samples, # class regularization samples, etc. I've read the paper, and found that they actually unfreeze everything including the text embedder (and VAE? I'm not sure so I leave it frozen), so I implemented textual inversion within the dreambooth example (new token, unfreeze a single row of the embedder), which improves results considerably, but the whole model still degrades no matter what.

Someone smarter than me can confirm, but I think the culprit is ADAM:

https://github.com/huggingface/diffusers/blob/f1484b81b0b763aa010fd2083df933305afb1812/examples/dreambooth/train_dreambooth.py#L426

My hypothesis is that since ADAM tries to drag all weights of unet etc. to 0, it ruins parts of the model that aren't concurrently being trained during the finetuning.

I've tested with `weight_decay` set to 0, and results seem considerably better, but I think the entire model is still degrading. I'm trying SGD next, so, fingers crossed, but there may still be some dragon lurking in the depths even despite removing ADAM.

A point of reference on this journey is the JoePenna "Dreambooth" library which doesn't implement PPL, and yet preserves priors much much better than this example, not to mention it learns the instance better, and is far more editable, and preserves out-of-class rather well. I expect more from this huggingface dreambooth example, and I'm trying to find why it's not delivering.

Any thoughts or guidance?

**EDIT1A**: SGD didn't learn the instance at 1000 steps + lr=5e-5, but it definitely preserved the priors way better (upon visual inspection. The loss really doesn't decrease much in any of my inversion/dreambooth experiments).

**EDIT1B**: Another test failed to learn using SGD at 1500 steps + lr=1e-3 + momentum=0.9. It might be trying to learn, but, not much. Priors were nicely preserved though still.

**EDIT1C**: 1500 * lr=5e2 learned great, was editable, didn't destroy other priors!!!

**EDIT2**: JoePenna [seems to use AdamW](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/main/ldm/models/diffusion/ddpm.py#L429), so I'm not sure what's up anymore, but I'm still getting quite poor results training with this library's (huggingface's) DB example.

**Comment by osanseviero** [comment]
cc @patil-suraj @apolinario

**Comment by affableroots** [comment]
If I understand correctly, [this paper](https://arxiv.org/pdf/2208.12242.pdf) says they finetune the entire model, although in my experiments I agree and have kept the VAE frozen.

![image](https://user-images.githubusercontent.com/114898764/193882644-69ef9105-f4da-4d49-a78b-05e9f68f2ec4.png)


-------------

RE JoePenna, quoted from his readme:

> We're [now] realizing the "regularization class" bits of this code have no effect, and that there is little to no prior preservation loss.

> So, out of respect to both the MIT team and the Google researchers, I'm renaming this fork to: "Unfrozen Model Textual Inversion for Stable Diffusion".

> For an alternate implementation that attempts prior loss preservation, please see ["Alternate Option"](https://github.com/JoePenna/Dreambooth-Stable-Diffusion#hugging-face-diffusers) below.

--------------

RE my params, I've tried ~50ish permutations within the following bounds, but in general I've tried to stick close to the example params in your readme.

* Batches: 1-2
* 8bit optimizer on/off
* Number of images: 5 - 50
* Number of prior images: 200-1500
* Number of steps: 500 - 5000
* Learning rate: 1e-7 to 1e-4
* Prior Preservation Loss on/off
* Prompts are always of the form: "a photo of CLASS" and "a photo of TOKEN CLASS", where TOKEN is `sks` (struggles to train away from rifles), or random three letter words, or, more recently, I added textual inversion to your example to add a token to the tokenizer.
* I've tried many types of images, but have settled on a test case of simply learning faces that aren't already in SD, eg, my face.

-----------------

RE degradation, I mean the whole model slowly becomes unusable over the course of learning. For instance given the stock params of 500 steps * lr=5e-6, it poorly learns the instance face, and all the class images visually degrade somewhat. It forgets things it knows already, and generally turns to visual noise, at any learning rate, and for any number of steps, in or out of the finetuning set. Very low learning rates, eg 1e-7, slow the process, but then you're not learning your instance either.

For inference I always test with: "a photo of token class", " a black and white sketch of token class", "a photo of johnny depp". So I've got in there: the exact prompt I trained with, a prompt that tests generalization, and a non-instance prompt to see if it's forgetting other faces. Johnny Depp degrades over the course of learning.

**Comment by jslegers** [comment]
> I'm still getting quite poor results training with this library's (huggingface's) DB example.

I've been using [ShivamShrirao's fork](https://github.com/ShivamShrirao/diffusers/blob/main/examples/dreambooth/DreamBooth_Stable_Diffusion.ipynb) of this repo for several days now. With [my own Google Colab notebook](https://colab.research.google.com/github/jslegers/diffusers/blob/main/examples/dreambooth/DreamBooth_Stable_Diffusion.ipynb) (which adds additional configuration options & streamlines the whole process), I've managed to generate some pretty decent renders of my younger self...

### Some of the renders generated
![image](https://user-images.githubusercontent.com/883647/193908596-c3a9c2f0-f052-4f83-baef-a2543add8795.png)

### Actual photos of me used as input for the training process
![image](https://user-images.githubusercontent.com/883647/193911401-189cfa7d-dfdb-46b1-b90d-b55a7d8b60fd.png)

**Comment by affableroots** [comment]
@jslegers Thanks for the follow up! I've had results look fairly similar, but I think everythings still degrading at least on my end. For instance:

* "a photo of" type prompts look a bit overcooked for the trained instance, as well as out-of class instances (unless my lr is low enough to not learn the instance)
* How do results look for random actors, compared to pre-training? Mine tend to degrade considerably, and often look "overcooked", ie that look things get at real high CFG.
* And how about using your likeness in non-portrait situations like, playing a sport, or an instrument, etc.?

**Comment by affableroots** [comment]
Oh, another symptom that I think indicates something is wrong, is I have not been able to **overtrain**. What I mean is, given a very high lr and/or high step count, everything sort of degrades without looking like my new instance. Faces look melted, and drift away from their priors, eg actors faces are no longer their faces, and they probably "melt" and become very indistinct, as faces. I notice photos degrade quicker than artwork.

Contrarily, I feel like JoePenna's version is easy to overtrain. Sure the trained concept leaks into everything, but I'm using this as a sort of diagnostic tool. In his version, an overtrained concept looks unmistakably like that concept (even if it increasingly fails to generalize)

**Comment by jslegers** [comment]
> but I think everythings still degrading at least on my end.

Fir my use case, I noticed that the diversity with the class "man" started declining significiantly. At some point, all random class images generated during the training process look identical. 

I haven't really looked much into the degrading of my "environment", though, as I was too focused to trying to fix my likeness. But it does indeed appear to be degrading as training increases.

> How do results look for random actors, compared to pre-training? Mine tend to degrade considerably, and often look "overcooked", ie that look things get at real high CFG.

Is this that "overcooked" look you're talking about?

Nightmare fuel, if you ask me...

![image](https://user-images.githubusercontent.com/883647/193966601-13b8ec1b-54c3-4516-b40e-f85e4e70c21c.png)

> And how about using your likeness in non-portrait situations like, playing a sport, or an instrument, etc.?

I haven't really invested much time in doing full body renders, but my best models did produce some OK results :

![image](https://user-images.githubusercontent.com/883647/193970045-911f9dbe-7fc1-4d7e-b9a3-b45e4cd3d704.png)

> he way I see it is if you have a whole universe of latent space, but you only update the small corner that looks like your face, but regardless you regularize the entire thing towards 0

That seems to be what's happening, yes. 

As a sidenote, I'm struggling to find a way to build upon previous trainings. When I have a general "johnslegers" concept, for example, I'd like to build upon that training be training eg. the concept "close-up or johnslegers" or "detailed portrait of johnslegers", as to refine the model further. But this doesn't seem possible. Attempts to achieve this by tinkering with my class prompt, instance prompt, class name or concept name generally leads to a loss of previous training achieved and serious degrading the quality of my models.

**Comment by affableroots** [comment]
Thank you for running that again! It seems we have similar results, though, your generalization attempts do look better than what I've been achieving, even given the stock params from the readme. And yes, that's the overcooking I'm talking about. That seems specifically like certain latent values are... clipping, if that's the right term.

> try to refine the model further

Ya imagine that's impossible if the entire latent space is getting regularized and forgetting everything it learned in its original training. I've turned off `weight_decay` and while it seems to help, the whole latent space gets degraded anyway. 

So then I thought maybe the Prior Preservation Loss was affecting things incorrectly, and I turned _that_ off. So, ostensibly it should become a run of the mill finetuning of just unet, I think. I would expect that training without PPL would simply, for instance, pull all faces toward your face, but not destroy everything. But, all faces somehow just get cooked!

I'm strongly suspecting weight decay is an issue _and_ some other hidden factor.

**Comment by jslegers** [comment]
> Ya imagine that's impossible if the entire latent space is getting regularized and forgetting everything it learned in its original training.

Probably. 

I guess I better just start all over with the original SD model and play around with different image files & config settings to reproduce at minimum the quality I achieved to far than continue tinkering with existing models...

**Comment by patil-suraj** [comment]
@affableroots In the paper by all layers they mean all layers of the unet model trained, the rest i.e the text encode is kept frozen.

Looking at the discussion and also from my experiments the model does run into overfitting and catastrophic forgetting. To me this seems likely as we finetune the whole model with few images. For now I don't know any way to avoid this, I am going to discuss this with the author and let you know if we find anything.

To sum it up, I'm not sure if the script is broken, it's rather that the model quickly overfits.

**Comment by affableroots** [comment]
I appreciate your attention on this @patil-suraj!

**Comment by jslegers** [comment]
> Looking at the discussion and also from my experiments the model does run into overfitting and catastrophic forgetting. To me this seems likely as we finetune the whole model with few images. For now I don't know any way to avoid this, I am going to discuss this with the author and let you know if we find anything.

I suspect overfitting issues and other related issues can be reduced by finetuning your number of training steps, class prompt, number of variety of class images generated, quality and variety of your input images, etc. 

I've been experimenting with various strategies : use just a handful of pics, use many pics, auto-generate the class picks, choose class pics myself, do only 250 training steps per run, do 2000 training steps each time, etc.

My results after just a single run from the original SD model vary significantly depending on various of these criteria and I have yet to determine any pattern.

We should develop some best practices on how to optimize the model we're creating while keeping degradation of the latent space minimal. If anyone should be able to provide some info we all seem to be missing, that's going to be the original author...

**Comment by affableroots** [comment]
While we're exploring the tuning of DB from all angles, I'll mention some recently decent results. I'm using my modified Dreambooth + Textual Inversion:

* 5 new tokens+embeddings
* 2000 steps 
* lr5e-5 for text embeddings
* lr5e-6 for unet
* weight decay=0.0
* prior loss weight=0.1
* # class images=1500

I will say that faces are still getting cooked, so, I'm going to play with this some more.

My theory is that if you combine DB + TI, you can ask TI to learn more strongly without risking damage to the rest of the model (because you're only learning brand new embeddings that don't cause other domains to forget), but still train unet to help  out a little, where TI is struggling.

**Comment by jslegers** [comment]
> I'm using my modified Dreambooth + Textual Inversion

Any chance you can make your modified script available for the rest of us, so we can test it ourselves?

**Comment by affableroots** [comment]
@jslegers Forgive the mess: https://gist.github.com/affableroots/a36a74287c8eb2da438a459795b158d6
I flew through it to clean it up and haven't tested again since, so, hopefully I didn't mess anything up but let me know if so!

This is TI + Unet unfreezing + Prior Preservation Loss. Text embeddings can learn at a different rate than unet. Hit me up with any questions, and good luck!

**Comment by affableroots** [comment]
@patil-suraj I know you're slammed, (I see you on every support ticket on this repo!), but any chance you've had an opportunity to ask the Dreambooth authors about the catastrophic forgetting + overfitting we've been facing? My pet theory was it has to do with weight decay, but, whatever it is, I'm just curious! No rush of course, and thanks for all your hard work!

**Comment by patrickvonplaten** [comment]
Gently pinging @patil-suraj here again :-)

**Comment by jslegers** [comment]
I did some testing regarding the impact of Dreambooth on different prompts, using the same seed.

Pretty much all of my tests produced results similar to this :

![afbeelding](https://user-images.githubusercontent.com/883647/194911616-d5e01a17-6a31-49d6-9ae3-5eb22318a85f.png)

**Comment by jslegers** [comment]
> did you have a chance to test that script that adds Textual Inversion to the mix?

Not yet. I'm planning to run some tests on Textual Inversion soon... both as a standalone approach, and using the script you provided.

I wanted to first complete my tests runs with Dreambooth alone, to see if I could find a way to reduce degradation by eg. increasing the number of class pics, reducing the prior loss weight or Adam weight decay, by using a different class (I tried "Jason Molina", who I allegedly kinda look like) or anything else I could think of. But nothing I tried produced better results. At least it allowed me to compile above overview, though, of the typical impact my test runs had.

**Comment by affableroots** [comment]
@jslegers Have you by chance tried the JoePenna repo? I'm still trying to pin down the difference, but, I think it works better, and I don't know why. 

They're starting to look identical, so I don't know where the difference in perceptual quality lies. The code is semantically identical, I think:

* cond stage (text) [unfrozen](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/main/configs/stable-diffusion/v1-finetune_unfrozen.yaml#L15).
* does use [PPL](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/5347a03a48828d1b54b5c6764f38293a985bdf3a/ldm/models/diffusion/ddpm.py#L921)
* VAE remains [frozen](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/main/ldm/models/diffusion/ddpm.py#L539). I thought this might not be the case because it "unfreezes the whole model", but, I think this does indeed stay frozen.
* regularization weight [= 1.0](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/5347a03a48828d1b54b5c6764f38293a985bdf3a/ldm/models/diffusion/ddpm.py#L447)
* AdamW [shares lr](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/5347a03a48828d1b54b5c6764f38293a985bdf3a/ldm/models/diffusion/ddpm.py#L1457)
* AdamW never sets `decay_weight`, leaving it to pytorches `1e-2` [default](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html#torch.optim.AdamW)

**Comment by patrickvonplaten** [comment]
Very interesting! @patil-suraj let's try this one out :-)

**Comment by affableroots** [comment]
Hm, I'm not sure I understand DDPM well yet, that timestep thing may be a red herring.

Could it be a difference of the DDPMP beta schedule type: diffuser's ["scaled_linear"](https://github.com/huggingface/diffusers/blob/24b8b5cf5e44b585e8808c34616cfa20e8e39866/examples/dreambooth/train_dreambooth.py#L418) vs JP's ["linear"](https://github.com/JoePenna/Dreambooth-Stable-Diffusion/blob/5347a03a48828d1b54b5c6764f38293a985bdf3a/ldm/models/diffusion/ddpm.py#L51)

If I throw enough mud at the wall, some will stick right?

**Comment by apolinario** [comment]
@affableroots, besides the loss looking smoother, what were your actual perceptual impressions from this experiment?

**Comment by patil-suraj** [comment]
Thanks a lot @affableroots and sorry for being later here. Running some experiments today, also comparing against other codebases. Will post my findings here soon.

**Comment by nerdyrodent** [comment]
I'm using Shivam's DB and I don't get that blotchy look unless I put the guidance scale above 30 during inference (or use too few steps on something like LMS). All the faces do start to tend towards my face though, which is nightmare fuel all of it's own! Tried with and without fp16 and varying amount of class images. I've also tried made up classes, such as §. It just seems to recognise a face as a face!

**Comment by matteoserva** [comment]
The blotchy look disappears if I increase the timesteps at inference time to a high value, like 250

**Comment by nerdyrodent** [comment]
Been testing using my face all day with a variety of optimisers because reasons :) Things learnt so far -
* It for sure changes the whole model regardless of "class"
* The defaults seem good so far, with 800-1800 steps being OK for a face
* My face bleeds into all the other faces, though some (Abraham Lincoln, Donald Trump) are impacted less

**Comment by compustar** [comment]
@nerdyrodent, have you tried 1e-6 for 2000 steps which is basically the JoePenna's default training parameters.

**Comment by EstudioAI** [comment]
**This "overcooked" look is actually the opposite, that happened to me after training for several steps (3000+) and trying to render using 50 inference steps or so.** If you look at this image, the early steps for some sampling methods look kinda similar to the bleeding we are talking about.

![2022101401](https://user-images.githubusercontent.com/115781692/195800829-21abdc50-eb6a-48b5-938a-b6c95ac8a44f.jpg)

I solved the issue:
- Increasing the inference steps (200+).
- Augmenting my images by flipping, rotating, and adding brightness, contrast or saturation.
- Adding more regularization images, in my case 180.

This way I can train more steps and reduce the influence over the rest of the model. I hope this info is useful, I can't promess it solves all the problem but I'm getting consistent results after 1000-2000 training steps.

![2022101402](https://user-images.githubusercontent.com/115781692/195804646-6d9e1442-fc2a-4f3a-b11b-7cd29c2fa380.jpg)

**Comment by nerdyrodent** [comment]
@compustar - just tried 2000 steps at 1e-6 and the results weren't as good as the default 800 steps at 5e-6.

**Comment by zcorley** [comment]
> 

Great thread you started here @affableroots 

I've been trying to run your [TI-Dreambooth-Tokenizer code](https://gist.github.com/affableroots/a36a74287c8eb2da438a459795b158d6). I am getting errors about accessing huggingface-cli that I am not getting when I run Shivam's version of the code. specifically:

OSError: CompVis/stable-diffusion-v1-4 is not a local folder and is not a valid model identifier listed on 'https://huggingface.co/models'
If this is a private repository, make sure to pass a token having permission to this repo with `use_auth_token` or log in with `huggingface-cli login`.
Traceback (most recent call last):

I am logged into hugface-cli tried; tried different tokens with no success.
I tried downloading the files locally and pathing them there, but eventually it asks for a config file i cannot find, however  this seems to be an access problem rather than a file location problem.

do you (or anyone) have an updated version, or do you know how i can fix this?

**Comment by d8ahazard** [comment]
> JoePenna "Dreambooth"

You need to download the *diffusers* model, which is a repo with multiple directories and files. The .ckpt is for regular stable diffusion, etc.

**Comment by patrickvonplaten** [comment]
Think @patil-suraj will very soon release a nice update of dreambooth :-)

**Comment by affableroots** [comment]
@patil-suraj Did you figure out what the issue was then? I've been curious.

**Comment by matteoserva** [comment]
It might be related to this PR: https://github.com/huggingface/diffusers/pull/883

**Comment by patil-suraj** [comment]
Hi everyone! Sorry to be so late here. 

We ran a lot of experiments with the script to see if there's any issues or if it's broken.
Turns out, we need to carefully pick hyperparameters like LR and training steps to get good results with dreambooth.

Also, the main reason the results with this script were not as good as Compvis forks, is that the `text_encoder` is being trained in those forks, and it makes a big difference in quality especially on faces.

We compiled all our experiments in this [report](https://wandb.ai/psuraj/dreambooth/reports/Dreambooth-training-analysis--VmlldzoyNzk0NDc3), and also added an option to train the `text_encoder` in the script which can be enabled by passing the `--train_text_encoder` argument.

Note that, if we train the `text_encoder` the training won't fit on 16GB GPU anymore, it will need at least 24GB VRAM. It should still be possible to do it 16GB using deepspeed, but will be slower.

Please take a look at the report, hope you find it useful.

**Comment by matteoserva** [comment]
I tried ShivamShirao's version and it trained correctly on a 12GB cpu. I don't know if it trained correctly the text encoder but I see that the weights of the text encoder have been updated. 
For instance, I used the adam8bit optimizer with gradient checkpointing and prior preservation loss

**Comment by matteoserva** [comment]
I tested ShivamShirao's fork with a 12GB card and it trained without errors, I seems that the text_encoder's weight have been updated. I haven't had time to compare the results of the two forks.
My parameters: with 8bit_adam, without caching latents, with prior preservation, without mixed precision, with gradient checkpointing, with text encoder training.

**Comment by d8ahazard** [comment]
I'll see if I can implement that one today.

**Comment by affableroots** [comment]
Thanks for the review @patil-suraj! It sounds like the difference was all in unfreezing the text embeddings. It sounds like lr+nsteps makes a small target to hit, and your experiment helps us aim better.

**Comment by jslegers** [comment]
Why is this issue closed?

I tried it out after the updates yesterday, with various different settings, but I barely noticed a difference.

Especially the overcooking of Johnny Depp keeps persisting, unless I lower the number of training steps low enouh the result for the trained subject no longer looks like me.

IMO the issue is far from fixed!

**Comment by 10pratik** [comment]
@affableroots I am getting good outputs for Dreambooth by training the images of a certain face. However I am not getting consistently good results i.e the model doesn't product the similar face. It changes it a little bit in some instances. 
Even when I take the best output and just change the prompt, it changes the face a bit.

Any idea on how to solve the issue?
